sharktank_models/llama3.1/test_llama.py

```diff
@@ -109,20 +95,56 @@ def decode(self, *, ids, start=0):
         return logits
 
     def allocate_cache(self):
-        return [self.to_device(numpy.zeros((128, page_size), numpy.float16))]
+        return [
+            self.to_device(
+                numpy.zeros((128, page_size // self.sharding), numpy.float16)
+            )
+            for i in range(self.sharding)
+        ]
 
     def to_device(self, a):
         return iree.runtime.asdevicearray(self.device, a)
 
 
-@pytest.fixture
-def toy_llama_cpu(compile_llama_cpu, llama_irpa):
-    return ToyLlama(compiled=compile_llama_cpu, device_id="local-task", irpa=llama_irpa)
+def cpu_flags(sharding):
+    return ["--iree-hal-target-device=llvm-cpu"] * sharding + [
+        "--iree-llvmcpu-target-cpu=host"
+    ]
 
 
-@pytest.fixture
-def toy_llama_hip(compile_llama_hip, llama_irpa):
-    return ToyLlama(compiled=compile_llama_hip, device_id="hip", irpa=llama_irpa)
+def hip_flags(sharding):
+    if "HIP_TARGET" not in os.environ:
+        raise RuntimeError("HIP_TARGET not set")
+
+    target_gpu = os.environ["HIP_TARGET"]
+    return ["--iree-hal-target-device=hip"] * sharding + [
```

Comment:
We would be running all shards on the same device. May not catch some problems where we mistakenly mix buffers on different devices where we don't expect DMA.
At some point may be worth running the test on a multi-device system.

Reply:
It should be good enough for base testing.